petals_local_parallel/all_to_all_communication_primitives.py

```diff
@@ -37,29 +44,30 @@ def __call__(self, x: torch.Tensor, rank: int):
             self.parts_ready[rank].clear()
 
 
-class AllGather:
-    def __init__(self, world_size: int, gather_op = torch.cat, ordered: bool = True):
+class AllGather(CollectiveOpetationBase):
+    def __init__(self, world_size: int, gather_op: callable = torch.cat):
         self.world_size = world_size
-        self.parts = []
+        self.parts: List[Optional[torch.Tensor]] = [None for _ in range(world_size)]
+        self.ranks_updated = []
         self.parts_ready = threading.Event()
         self.gather_op = gather_op
-        self.ordered = ordered
-        
+
     def __call__(self, x: torch.Tensor, rank: int):
+        parts, ranks_updated, parts_ready = self.parts, self.ranks_updated, self.parts_ready
+        # ^-- note: we copy properties to locals so that the "finally" clause is thread-safe
         try:
-            parts, parts_ready = self.parts, self.parts_ready
-            # ^-- note: we copy properties to locals so that the "finally" clause is thread-safe
-            parts.append((rank, x))  # append is thread-safe. thanks, GIL!
-            if len(parts) == self.world_size:
-                parts_ready.set() # can be called more than once; we dont care
+            parts[rank] = x  # no race b/c each rank writes to a separate location
+            ranks_updated.append(rank)  # append is thread-safe. thanks, GIL!
+            if len(ranks_updated) == self.world_size:
+                parts_ready.set()  # can be called more than once; we dont care
             parts_ready.wait()
-            if self.ordered:
-                parts = sorted(parts)  # sorted by rank
-            parts = [part.to(x.device, non_blocking=True) for r, part in parts]
+            parts = [part.to(x.device, non_blocking=True) for part in parts]
             # note: for one of the parts with r == rank, part.to(device) is a no-op
-            return self.gather_op(parts, dim=-1)
+            return self.gather_op(parts)
```

Comment:
rationale: AllGather is a generic op that can accept arbitrary gather_op callable; hardcoding dim here makes it hard-coded to torch.cat. Instead, we can set dim=-1 when instantiating AllGather (see slicer_wrapper.py diff)